PR: Add current version of docs style guide

CAM-Gerlach opened the pull request:
<!--- Before submitting your pull request, --->
<!--- please complete as much as possible of the following checklist: --->

### Pull Request Checklist

* [x] Read and followed this repo's [Contributing Guidelines](https://github.com/spyder-ide/spyder/blob/master/CONTRIBUTING.md)
* [x] Based your PR on the latest version of the correct branch (master or 3.x)
* [x] Checked your writing carefully for correct English spelling, grammar, etc
* [x] Described your changes and the motivation for them below
* [x] ~~Noted what issue(s) this pull request resolves, creating one if needed~~



## Description of Changes

<!--- Describe what you've changed and why. --->

Add the styleguide I originally created and developed during the big docs revamp and that all the current docs follow. If the current maintainers wish, further changes can be made but this should be committed in its initial state (plus any corrections/any non-substantive changes) to represent the current state of the docs as they are now, with any further changes then tracked seperately.


<!--- Thanks for your help making Spyder --->
<!--- and its documentation better for everyone! --->

CAM-Gerlach commented:
Since I can't tag people for review anymore, @ccordoba12 @juanis2112 .

juanis2112 commented:
Thanks for this @CAM-Gerlach , I'll take a look at it.

CAM-Gerlach commented:
@juanis2112 I suggest you merge this as a starting unless there are any obvious mistakes, corrections and the like, since it represents the style guide as actually implemented in the docs as they are now, and then making any further additions, changes and removals in a separate PR so its clear what was changed and will be changed in the future docs.

ccordoba12 commented on STYLEGUIDE.md:
```diff
@@ -0,0 +1,101 @@
+
+## reST Style Elements
+
+- Filenames/paths: Use :file: and ``/`` for deliminators e.g. :file:`path/to/test.py`. Example filenames should be all lowercase with no spaces or special characters unless called for by the application; real filenames should follow their actual case. Use {descriptive-name} for placeholders.
```
Is `deliminators` a word, or you meant `delimiters` here?

CAM-Gerlach replied:
```suggestion
- Filenames/paths: Use :file: and ``/`` for delimiters e.g. :file:`path/to/test.py`. Example filenames should be all lowercase with no spaces or special characters unless called for by the application; real filenames should follow their actual case. Use {descriptive-name} for placeholders.
```

"deliminator" actually is a word, but delimiter was what I meant. Good catch, thanks.

ccordoba12 commented on STYLEGUIDE.md:
```diff
@@ -0,0 +1,101 @@
+## reST Style Elements
+
+- Filenames/paths: Use :file: and ``/`` for deliminators e.g. :file:`path/to/test.py`. Example filenames should be all lowercase with no spaces or special characters unless called for by the application; real filenames should follow their actual case. Use {descriptive-name} for placeholders.
+- Keyboard Shortcuts: Use :kbd: to format, Ctrl/Alt/Shift as abbreviations (instead of Command/Option), ``-`` with no spaces to separate, and title case (including capital letters for letter keys; always include Shift explicitly if needed) e.g. :kbd:`Ctrl-Alt-Shift-P`. Use the key name or symbol as printed on a standard keyboard. For any configurable shortcut, you should include "by default" to i ...
```
another -> other ?

CAM-Gerlach replied:
```suggestion
- Keyboard Shortcuts: Use :kbd: to format, Ctrl/Alt/Shift as abbreviations (instead of Command/Option), ``-`` with no spaces to separate, and title case (including capital letters for letter keys; always include Shift explicitly if needed) e.g. :kbd:`Ctrl-Alt-Shift-P`. Use the key name or symbol as printed on a standard keyboard. For any configurable shortcut, you should include "by default" to indicate this, and must include at least one other means the command can be accessed if available (e.g. via a menu or toolbar), which should be done for all shortcuts and should include all means of executing the command, if multiple others exist.
```

Yup, it actually took me a minute to figure out that one myself, heh.

ccordoba12 commented on STYLEGUIDE.md:
````diff
@@ -0,0 +1,101 @@
+
+- Filenames/paths: Use :file: and ``/`` for deliminators e.g. :file:`path/to/test.py`. Example filenames should be all lowercase with no spaces or special characters unless called for by the application; real filenames should follow their actual case. Use {descriptive-name} for placeholders.
+- Keyboard Shortcuts: Use :kbd: to format, Ctrl/Alt/Shift as abbreviations (instead of Command/Option), ``-`` with no spaces to separate, and title case (including capital letters for letter keys; always include Shift explicitly if needed) e.g. :kbd:`Ctrl-Alt-Shift-P`. Use the key name or symbol as printed on a standard keyboard. For any configurable shortcut, you should include "by default" to i ...
+- GUI elements (names of things in the UI): Use ``:guilabel:`` , e.g. ``:guilabel:`Name```; make sure to use the exact name and capitalization as printed in the UI, except for the proper names of Spyder panes (i.e. those listed in the ``View > Panes`` menu), which should always be written in Title Case ("History Log", "Project Explorer", "Find in Files", etc. as they are proper names). Also, for  ...
````
* There are three backticks at the end of this expression:

      ``:guilabel:`Name```

* "History Log", "Project Explorer", "Find in Files": these examples don't apply anymore. Please change them to "Variable Explorer", "IPython Console".

CAM-Gerlach replied:
```suggestion
- GUI elements (names of things in the UI): Use ``:guilabel:`` , e.g. ``:guilabel:`Name` ``; make sure to use the exact name and capitalization as printed in the UI, except for the proper names of Spyder panes (i.e. those listed in the ``View > Panes`` menu), which should always be written in Title Case ("Variable Explorer", "Code Analysis", "IPython Console", etc. as they are proper names). Also, for consistency and clarity, they should be referred to as panes, rather than modules, plugins, panels, etc.; Find in Files may be called a "tool" if appropriate to the context.
```

> There are three backticks at the end of this expression:

Yes, which is the correct number. The first is the closing backtick for `Name`, and the second two are the closing backticks for the inline verbatim text. Apparently Github's GFM parser doesn't render this correctly so I've added a space to fix that.

> Please change them to "Variable Explorer", "IPython Console".

Fixed to three up to date examples. Ironically, I just noticed that in the Spyder UI, two of those I mentioned are Sentence case, while one is Title Case, and the rest of the UI text isn't much better. This is really my fault for never doing that polish pass I was going to...I can get back to it some day.

ccordoba12 commented on STYLEGUIDE.md:
````diff
@@ -0,0 +1,101 @@
+- Filenames/paths: Use :file: and ``/`` for deliminators e.g. :file:`path/to/test.py`. Example filenames should be all lowercase with no spaces or special characters unless called for by the application; real filenames should follow their actual case. Use {descriptive-name} for placeholders.
+- Keyboard Shortcuts: Use :kbd: to format, Ctrl/Alt/Shift as abbreviations (instead of Command/Option), ``-`` with no spaces to separate, and title case (including capital letters for letter keys; always include Shift explicitly if needed) e.g. :kbd:`Ctrl-Alt-Shift-P`. Use the key name or symbol as printed on a standard keyboard. For any configurable shortcut, you should include "by default" to i ...
+- GUI elements (names of things in the UI): Use ``:guilabel:`` , e.g. ``:guilabel:`Name```; make sure to use the exact name and capitalization as printed in the UI, except for the proper names of Spyder panes (i.e. those listed in the ``View > Panes`` menu), which should always be written in Title Case ("History Log", "Project Explorer", "Find in Files", etc. as they are proper names). Also, for  ...
+- Menu items and preference panes: Use :menuselection: to format and ``-->`` to separate, e.g. `` :menuselection:`Project --> New Project` ``, make sure to use the exact name and capitalization of each level element
````
* Why not use `>` instead of `-->` here? That's the convention I've been following for years now.
* Also, I think you should use a dot instead of a comma here:

      :menuselection:`Project --> New Project` ``. Make sure ...

CAM-Gerlach replied:
Because `-->` is the correct delimiter for `:menuselection:` and `>` will not work, per the [Sphinx docs](https://www.sphinx-doc.org/en/master/usage/restructuredtext/roles.html#role-menuselection). And also, again, it is what is implemented across all the docs; any changes should be made as a further PR that changes it both here and there.

CAM-Gerlach replied:
> Also, I think you should use a dot [sic] instead of a comma here:

Correct, you've detected a comma splice. Also FYI, in English its called a "period" in this context (in British English, it would be a "full stop").

```suggestion
- Menu items and preference panes: Use :menuselection: to format and ``-->`` to separate, e.g. `` :menuselection:`Project --> New Project` ``. Make sure to use the exact name and capitalization of each level element
```

ccordoba12 commented on STYLEGUIDE.md:
````diff
@@ -0,0 +1,101 @@
+- GUI elements (names of things in the UI): Use ``:guilabel:`` , e.g. ``:guilabel:`Name```; make sure to use the exact name and capitalization as printed in the UI, except for the proper names of Spyder panes (i.e. those listed in the ``View > Panes`` menu), which should always be written in Title Case ("History Log", "Project Explorer", "Find in Files", etc. as they are proper names). Also, for  ...
+- Menu items and preference panes: Use :menuselection: to format and ``-->`` to separate, e.g. `` :menuselection:`Project --> New Project` ``, make sure to use the exact name and capitalization of each level element
+- Short commands/lines of code (under 20-30 characters), use `` , e.g. ``code``
+- For any code longer than one line, ~20-30 characters, or that is otherwise important to emphasize rather than just mention, use ``code-block::`` tags, closed with ``end``, and specify the programming language (bash for command line, python for python, ini for configuration files, etc.). PEP 8 should be followed for Python code, with all "recommended"/"preferred"/"suggested" guidelines being tre ...
````
Python ~~3.4-3.6~~ 3.5-3.7

CAM-Gerlach replied:
Does Spyder still not support 3.8 yet?

ccordoba12 commented on STYLEGUIDE.md:
```diff
@@ -0,0 +1,101 @@
+- Short commands/lines of code (under 20-30 characters), use `` , e.g. ``code``
+- For any code longer than one line, ~20-30 characters, or that is otherwise important to emphasize rather than just mention, use ``code-block::`` tags, closed with ``end``, and specify the programming language (bash for command line, python for python, ini for configuration files, etc.). PEP 8 should be followed for Python code, with all "recommended"/"preferred"/"suggested" guidelines being tre ...
+- Internal links to sections should never rely on implicit references, since they may change with section titles or other re-organization. Instead, always use explicit references above the target with one blank line between; keep ref names short, use ``-`` to replace spaces, use unique and distinctive names and append ``-ref`` to the end of the name for clarity.
+- All external links should be inline (not bare) unless specifically needed to illustrate some property of the link itself, should be https if available, and should, generally, be written as a reference with `` `Test Link Name`_ `` and then included in full at after the end of the paragraph in which it appears, e.g. `` .. _Test Link Name: https://www.example.com/ `` . A block of links should be s ...
```
* ... then included in full ~~at~~ after the end ...
* Make sure to check that your links <something text seems to be missing here?>, as the build will automatically check for broken ones.
* I don't understand what you mean by this: "Strip anything unnecessary (parameters, referrers, etc) off the end. As the base docs are in US English, any links should be to the English, desktop version of the site."

CAM-Gerlach replied:
```suggestion
- All external links should be inline (not bare) unless specifically needed to illustrate some property of the link itself, should be HTTPS if available, and should, generally, be written as a reference with `` `Test Link Name`_ `` and then included in full after the end of the paragraph in which it appears, e.g. `` .. _Test Link Name: https://www.example.com/ `` . A block of links should be separated from paragraphs above and below by one blank line, with no blank lines between links. Make sure to check your links, as the build will automatically check for broken ones. Strip anything unnecessary (parameters, referrers, etc) off the end. As the base docs are in US English, any links should be to the English, desktop version of the site.
```

Fixed and fixed.

E.g. `example.com/page?tracking=fdsds&crap=fdssds&here=dsllds` -> `example.com/page`. Takes up less space, is less invasive of user privacy, looks cleaner and more trustworthy, and avoids any unintentional side effects.

ccordoba12 commented on STYLEGUIDE.md:
```diff
@@ -0,0 +1,101 @@
+
+## Images/Figures
+
+- Organize images into directories following the Google drive
```
What do you mean by this?

CAM-Gerlach replied:
```suggestion
```

Its out of date now since the huge trache of screenshots I made is now obsolete with Spyder 4's release.

ccordoba12 commented on STYLEGUIDE.md:
```diff
@@ -0,0 +1,101 @@
+
+- Organize images into directories following the Google drive
+- Lower case file names and extensions, separate with hyphens, 3-character extensions
+- For ease of visibility, explanatory images should be full width; expository or those with very simple subjects can be half-width, right aligned
```
Why right aligned? Shouldn't they be left aligned?

CAM-Gerlach replied:
There may be another technical reason, but the layout flows better that way, and it makes conceptual sense for secondary, less important images to be such as they are supplementary and illustrative of the text, rather than the primary focus (or they'd be full width). That, and as I've emphasized multiple times its what is currently implemented in the docs; if you want to change it you can make an atomic PR after this one that changes it both here and in the docs themselves.

ccordoba12 commented on STYLEGUIDE.md:
```diff
@@ -0,0 +1,101 @@
+
+- Follow the [OpenStack General Writing Guidelines](https://docs.openstack.org/doc-contrib-guide/writing-style/general-writing-guidelines.html)
+- Maintain an explanatory rather than expository tone if appropriate for the language (i.e. in English, "How" rather than "What"; e.g. "Spyder can do XXX." -> "To do XXX, you can...")
+- Focus on the user; this is documentation, not promotional material
```
Good point!

CAM-Gerlach replied:
I sometimes forget this

ccordoba12 commented on STYLEGUIDE.md:
```diff
@@ -0,0 +1,101 @@
+- Maintain an explanatory rather than expository tone if appropriate for the language (i.e. in English, "How" rather than "What"; e.g. "Spyder can do XXX." -> "To do XXX, you can...")
+- Focus on the user; this is documentation, not promotional material
+- Try to use simple language if possible and avoid buzzwords, jargon and acronyms
+- Make sure capitalization, spacing etc. of proper names and acronyms are correct and consistent per the original (or the canonical localized form is availible)
```
* I don't understand what you mean by this.
* is available -> if available ?

CAM-Gerlach replied:
```suggestion
- Make sure capitalization, spacing etc. of proper names and acronyms are correct and consistent per the original (or the canonical localized form if available)
```

Yup, fixed.

Your English language powers have doubled since we last met xD

ccordoba12 commented on STYLEGUIDE.md:
```diff
@@ -0,0 +1,101 @@
+## Writing (English)
+
+- Standard written English; US spellings are nominally the default for consistency but not required, and any given page should keep to only one convention. Avoid dialect-specific words.
+- AP style unless noted elsewhere here
```
What does this mean?

CAM-Gerlach replied:
Use AP style, unless stated otherwise by this style guide

CAM-Gerlach commented on STYLEGUIDE.md:
```diff
@@ -0,0 +1,101 @@
+
+## General
+
+Must (not), should (not), recommended, suggested, may, and can generally follow IETF conventions, with some additional differentiation:
```
https://ietf.org/

CAM-Gerlach replied:
All internet standards (RFCs) follow this along with numerous others.

CAM-Gerlach commented on STYLEGUIDE.md:
```diff
@@ -0,0 +1,101 @@
+- (Not) **Recommended** and **Preferred** clauses should be followed where practicable; a brief rationale should be provided for not doing so
+- **Suggested** clauses indicate a convention, approach or guideline that may be helpful to follow, but has no binding effect or justification requirements
+- **May** or **Can** clauses designate explicitly permitted guidelines, actions or conventions, often for clarity, that are allowed but have no weight of preference on them (and may even be gently discouraged with a suggested or preferred alternative)
+
```
Its just any clause that uses the term, it can be anything. The style guide currently doesn't actually use them thoroughly because I never did the polish pass I planned to long ago before submitting this. Since this section is only normative to the style guide itself rather than to the actual docs, and not consistently employed as of now, I could just remove it.

CAM-Gerlach commented on STYLEGUIDE.md:
```diff
@@ -0,0 +1,101 @@
+
+- In general, follow the [OpenStack Heading Guidelines](https://docs.openstack.org/doc-contrib-guide/writing-style/general-writing-guidelines.html) with regard to sections and headings; in particular, section and subsection titles delineating overall categories of tasks/topics and be in gerund (i.e. with "ing", e.g. "Editing code", "Debugging files"), while those concerning a concept or reference ...
+- Heading case: Page titles (L1 headings) should use Title Case, all others should use Sentence case
+- 1 L1 Heading for page titles: Use ``######`` with overline; on first line of file; keep to a few words
```
```suggestion
- One L1 Heading, used for page titles: Use ``######`` with overline; on first line of file; keep to a few words
```

Its to make explicit that there should only be one per document, but I've rewritten it to be clearer.

CAM-Gerlach commented on STYLEGUIDE.md:
```diff
@@ -0,0 +1,101 @@
+- Heading case: Page titles (L1 headings) should use Title Case, all others should use Sentence case
+- 1 L1 Heading for page titles: Use ``######`` with overline; on first line of file; keep to a few words
+- L2 Headings for section heads: Use ========= with overline; 3 blank lines before them, will appear in sidebar so keep them short
+- L3 headings for subheads: Use ~~~~~~, 2 blank lines,
```
```suggestion
- L3 headings for subheads: Use ~~~~~~, 2 blank lines
```

Good catch. There is a lot of sloppiness here since I never did the big polish pass I was going to.

CAM-Gerlach commented on STYLEGUIDE.md:
```diff
@@ -0,0 +1,101 @@
+- 1 L1 Heading for page titles: Use ``######`` with overline; on first line of file; keep to a few words
+- L2 Headings for section heads: Use ========= with overline; 3 blank lines before them, will appear in sidebar so keep them short
+- L3 headings for subheads: Use ~~~~~~, 2 blank lines,
+- L4 Headings for sub-subheads if truly necessary: Use ---------, 2 blank lines, keep to a minimum (consider a flatter structure or breaking into multiple files)
```
No, flatten would be a grammar error here, while flatter is correct. The former is an adjective (meaning "more flat") while the latter is a verb.

CAM-Gerlach commented on STYLEGUIDE.md:
```diff
@@ -0,0 +1,101 @@
+
+- Markdown, Github-flavored (``.md``) for any non-documentation rendered text files in the root of the repo, e.g. ``README``, ``CONTRIBUTING``, etc.
+- reStructuredText, (reST, ``.rst``) for documentation format, i.e. any user-visible files in the ``/doc`` directory
+- Where not otherwise contradicted by this guide or specific individual requirements, a subset of Markdown corresponding to that cross-compatible with reST should be preferred where possible; e.g. ``*italic*`` vs. ``_italic_``, ``*`` for bulleted lists, double backtick for inline code vs. single, etc.
```
I've addressed this before, at some length. For Markdown in the Spyder repo it may be justifiable since it is less likely to have to interoperate with the reST docs, Markdown is the common format, you have to edit it frequently and edits are much more short-form. However, given none of this is the case (and noting that on an US English keyboard layout, which would be expected to be used/switched to when writing the long-form English prose on this repo, the additional time spent is minuscule relative to the substantial overhead converting back and forth if needed, even with tools), I judged it to be worthwhile and the content of this repo is all formatted accordingly.

If you and @juanis2112 disagree and want to change it, that's fine, but the style guide is meant to reflect the current state of the docs and repo, where this is the case. Once this is committed, if you want to change it, it can then be changed repo-wide and in the style guide, and the changes submitted as an atomic PR, so its clear.

CAM-Gerlach commented on STYLEGUIDE.md:
```diff
@@ -0,0 +1,101 @@
+- Markdown, Github-flavored (``.md``) for any non-documentation rendered text files in the root of the repo, e.g. ``README``, ``CONTRIBUTING``, etc.
+- reStructuredText, (reST, ``.rst``) for documentation format, i.e. any user-visible files in the ``/doc`` directory
+- Where not otherwise contradicted by this guide or specific individual requirements, a subset of Markdown corresponding to that cross-compatible with reST should be preferred where possible; e.g. ``*italic*`` vs. ``_italic_``, ``*`` for bulleted lists, double backtick for inline code vs. single, etc.
+- Documentation should build with no warnings with ``Sphinx`` in nitpicky mode
```
I did add it to Travis, it builds the docs in nitpicky mode and fails if so. However, that's all buried in a config file, putting it here makes it normative, persistent, and serves as a reminder for users to ensure their changes build.

CAM-Gerlach commented on STYLEGUIDE.md:
```diff
@@ -0,0 +1,101 @@
+- Lower case filenames, rst extension, hyphens as space deliminators, short but descriptive
+- End files with a newline (blank line)
+- 1 blank line after all headings and before and after paragraphs, directives and "|"s
+- Use Line Feed (LF, ``\n``) as the End of Line character on all platforms
```
Yes, of course. Again, like the above, this makes it normative, persistent and avoids any confusion over why line endings were normalized.

CAM-Gerlach commented on STYLEGUIDE.md:
```diff
@@ -0,0 +1,101 @@
+- 1 blank line after all headings and before and after paragraphs, directives and "|"s
+- Use Line Feed (LF, ``\n``) as the End of Line character on all platforms
+- Use UTF-8 for all files (with no BOM, as dictated by the UTF-8 standard); do not use Windows Notepad as this adds a BOM and can otherwise corrupt files easily
+- Indicate special characters (e.g. em dash, en dash, etc) by simply including them inline with Unicode
```
Instead of using escape sequences, simply use the actual UTF-8 character

CAM-Gerlach replied:
```suggestion
- Use standard UTF-8 characters instead of escape sequences for non-ASCII values (e.g. em dash, en dash, etc)
```

CAM-Gerlach commented on STYLEGUIDE.md:
```diff
@@ -0,0 +1,101 @@
+- Indicate special characters (e.g. em dash, en dash, etc) by simply including them inline with Unicode
+- Suggested to avoid non-ASCII characters where minimal downside/need exists, and recommended to keep used characters within the Latin-1 set where an additional range is not required by the specific usage, or an existing proper name or other identifier uses them
+- Spaces, no tabs for indentation and otherwise
+- Natural indents (3 spaces) for directives, 4 spaces for further indents beyond that in code
```
Because that's the "natural" visual indent level and matches most other sources and style guides. It can be changed later, but its what's implemented now repo-wide.

CAM-Gerlach commented on STYLEGUIDE.md:
```diff
@@ -0,0 +1,101 @@
+    - Wrap code blocks at a width of 70 characters from the page margin for readability.
+- Place section links above the section with one blank line between them and the section, and two blank lines between sections (as appropriate for that section level) between the link and the preceding text.
+- Always use single backticks around link titles
+- Eliminate any trailing whitespace before commiting
```
Not without [some nasty hacks](https://stackoverflow.com/questions/591923/make-git-automatically-remove-trailing-whitespace-before-committing). A much better solution is pre-commit hooks to enforce a lot of this stuff, but we'd have to get them set up and ensure everyone uses them. We could use [pre-commit](https://pre-commit.com/) which seems like a pretty nice and comprehensive framework for them, though someone would actually have to implement it.

CAM-Gerlach replied:
Right now I just have a bunch of small bash scripts that checks most of these things, but they are pretty horribly written and take some time to run on larger codebases.

CAM-Gerlach commented:
@ccordoba12 @juanis2112

CAM-Gerlach commented on STYLEGUIDE.md:
```diff
@@ -0,0 +1,101 @@
+## Structure
+
+- In general, follow the [OpenStack Heading Guidelines](https://docs.openstack.org/doc-contrib-guide/writing-style/general-writing-guidelines.html) with regard to sections and headings; in particular, section and subsection titles delineating overall categories of tasks/topics and be in gerund (i.e. with "ing", e.g. "Editing code", "Debugging files"), while those concerning a concept or reference ...
+- Heading case: Page titles (L1 headings) should use Title Case, all others should use Sentence case
```
@juanis2112 ^

CAM-Gerlach commented:
@ccordoba12 Hey, can you give this a final check so we can merge and @juanis2112 can make any changes she wants on top of this? Thanks.

CAM-Gerlach commented:
@ccordoba12

CAM-Gerlach commented:
I'm just going to go ahead and merge this per discussion with @juanis2112 , since I've addressed all the comments, and it represents the current state of the docs and what we're actively following so far, and it has waited 6 months for a review. You all are welcome to modify it as needed from here. Thanks!
